commit ae0f0b2aeb8bd196fd1e47790e3e99eb77d0feee (refs/changes/63/1163/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-12-21T19:03:07-08:00 (2 years, 10 months ago)
    
    joyent/node-cueball#47 rapid add/remove of backend in DNS leads to ConnectionSet crash

diff --git a/lib/resolver.js b/lib/resolver.js
index cf15e6e..840d67a 100644
--- a/lib/resolver.js
+++ b/lib/resolver.js
@@ -877,6 +877,10 @@ CueBallDNSResolver.prototype.state_process = function (S) {
 	}
 
 	this.emit('updated');
+
+	/* Write down what we did to help debugging. */
+	this.r_lastProcessed = { added: added, removed: removed };
+
 	S.gotoState('sleep');
 };
 
diff --git a/lib/set.js b/lib/set.js
index daf6366..d949ddc 100644
--- a/lib/set.js
+++ b/lib/set.js
@@ -141,9 +141,11 @@ CueBallConnectionSet.prototype.on_resolver_removed = function (k) {
 	});
 	delete (this.cs_dead[k]);
 	var fsms = self.cs_fsms[k];
-	mod_assert.ok(fsms.length <= 1);
-	if (cks.length === 0 && fsms[0] && !fsms[0].isInState('idle'))
-		fsms[0].close();
+	if (fsms !== undefined) {
+		mod_assert.ok(fsms.length <= 1);
+		if (cks.length === 0 && fsms[0] && !fsms[0].isInState('idle'))
+			fsms[0].close();
+	}
 };
 
 CueBallConnectionSet.prototype.isDeclaredDead = function (backend) {
